src/data/list/basic.lean

```diff
@@ -1047,6 +1045,72 @@ end
 
 end insert_nth
 
+/- map -/
+
+lemma map_congr {f g : α → β} : ∀ {l : list α}, (∀ x ∈ l, f x = g x) → map f l = map g l
+| []     _ := rfl
+| (a::l) h := let ⟨h₁, h₂⟩ := forall_mem_cons.1 h in
+  by rw [map, map, h₁, map_congr h₂]
+
+lemma map_eq_map_iff {f g : α → β} {l : list α} : map f l = map g l ↔ (∀ x ∈ l, f x = g x) :=
+begin
+  refine ⟨_, map_congr⟩, intros h x hx,
+  rw [mem_iff_nth_le] at hx, rcases hx with ⟨n, hn, rfl⟩,
+  rw [nth_le_map_rev f, nth_le_map_rev g], congr, exact h
+end
+
+theorem map_concat (f : α → β) (a : α) (l : list α) : map f (concat l a) = concat (map f l) (f a) :=
+by induction l; [refl, simp only [*, concat_eq_append, cons_append, map, map_append]]; split; refl
+
+theorem map_id' {f : α → α} (h : ∀ x, f x = x) (l : list α) : map f l = l :=
+by induction l; [refl, simp only [*, map]]; split; refl
+
+@[simp] theorem foldl_map (g : β → γ) (f : α → γ → α) (a : α) (l : list β) : foldl f a (map g l) = foldl (λx y, f x (g y)) a l :=
+by revert a; induction l; intros; [refl, simp only [*, map, foldl]]
+
+@[simp] theorem foldr_map (g : β → γ) (f : γ → α → α) (a : α) (l : list β) : foldr f a (map g l) = foldr (f ∘ g) a l :=
+by revert a; induction l; intros; [refl, simp only [*, map, foldr]]
+
+theorem foldl_hom (f : α → β) (g : α → γ → α) (g' : β → γ → β) (a : α)
+  (h : ∀a x, f (g a x) = g' (f a) x) (l : list γ) : f (foldl g a l) = foldl g' (f a) l :=
+by revert a; induction l; intros; [refl, simp only [*, foldl]]
+
+theorem foldr_hom (f : α → β) (g : γ → α → α) (g' : γ → β → β) (a : α)
+  (h : ∀x a, f (g x a) = g' x (f a)) (l : list γ) : f (foldr g a l) = foldr g' (f a) l :=
+by revert a; induction l; intros; [refl, simp only [*, foldr]]
+
+theorem eq_nil_of_map_eq_nil {f : α → β} {l : list α} (h : map f l = nil) : l = nil :=
+eq_nil_of_length_eq_zero $ by rw [← length_map f l, h]; refl
+
+@[simp] theorem map_join (f : α → β) (L : list (list α)) :
```

Comment:
Is this really a `simp` lemma? If anything I would have picked the reverse direction.

Reply:
Almost all of the lemmas about `map` were already there. The only one I added is `map_subset_iff `. I moved the others down, because `map_subset_iff` is using lemmas about `nth` that were below the section on `map` before.

Reply:
Okay!